In [1]:
import numpy as np
import pandas as pd
from sklearn . preprocessing import LabelEncoder , MinMaxScaler

In [2]:
names = ['duration', 'protocol', 'service', 'flag', 'src_bytes', 'dst_bytes', 'land','wrong_fragment', 'urgent', 'hot',
         'num_failed_logins', 'logged_in', 'num_compromised', 'root_shell', 'su_attempted', 'num_root',
         'num_file_creations', 'num_shells', 'num_access_files', 'num_outbound_cmds', 'is_host_login',
         'is_guest_login', 'count', 'srv_count', 'serror_rate', 'srv_serror_rate', 'rerror_rate', 'srv_rerror_rate',
         'same_srv_rate', 'diff_srv_rate', 'srv_diff_host_rate', ' dst_host_count', 'dst_host_srv_count',
         'dst_host_same_srv_rate', 'dst_host_diff_srv_rate', ' dst_host_same_src_port_rate', 'dst_host_srv_diff_host_rate',
         'dst_host_serror_rate', 'dst_host_srv_serror_rate', 'dst_host_rerror_rate', 'dst_host_srv_rerror_rate',
         'attack_type', 'other']

In [3]:
df = pd. read_csv ('KDDTrain+.txt', names =names , header = None )
dft = pd. read_csv ('KDDTest+.txt', names =names , header = None )
print ("Initial test and training data shapes :", df.shape , dft . shape )

Initial test and training data shapes : (125973, 43) (22544, 43)


In [4]:
full = pd. concat ([df , dft ])
assert full . shape [0] == df. shape [0] + dft. shape [0]

In [5]:
full.head()

duration protocol   service flag  src_bytes  dst_bytes  land  \
0         0      tcp  ftp_data   SF        491          0     0   
1         0      udp     other   SF        146          0     0   
2         0      tcp   private   S0          0          0     0   
3         0      tcp      http   SF        232       8153     0   
4         0      tcp      http   SF        199        420     0   

   wrong_fragment  urgent  hot  ...    dst_host_same_srv_rate  \
0               0       0    0  ...                      0.17   
1               0       0    0  ...                      0.00   
2               0       0    0  ...                      0.10   
3               0       0    0  ...                      1.00   
4               0       0    0  ...                      1.00   

   dst_host_diff_srv_rate   dst_host_same_src_port_rate  \
0                    0.03                          0.17   
1                    0.60                          0.88   
2                    0.05                          0.00   
3                    0.00                          0.03   
4                    0.00                          0.00   

   dst_host_srv_diff_host_rate  dst_host_serror_rate  \
0                         0.00                  0.00   
1                         0.00                  0.00   
2                         0.00                  1.00   
3                         0.04                  0.03   
4                         0.00                  0.00   

   dst_host_srv_serror_rate  dst_host_rerror_rate  dst_host_srv_rerror_rate  \
0                      0.00                  0.05                      0.00   
1                      0.00                  0.00                      0.00   
2                      1.00                  0.00                      0.00   
3                      0.01                  0.00                      0.01   
4                      0.00                  0.00                      0.00   

   attack_type  other  
0       normal     20  
1       normal     15  
2      neptune     19  
3       normal     21  
4       normal     21  

[5 rows x 43 columns]

In [6]:
full ['label'] = full ['attack_type']

In [7]:
full.head()

duration protocol   service flag  src_bytes  dst_bytes  land  \
0         0      tcp  ftp_data   SF        491          0     0   
1         0      udp     other   SF        146          0     0   
2         0      tcp   private   S0          0          0     0   
3         0      tcp      http   SF        232       8153     0   
4         0      tcp      http   SF        199        420     0   

   wrong_fragment  urgent  hot   ...     dst_host_diff_srv_rate  \
0               0       0    0   ...                       0.03   
1               0       0    0   ...                       0.60   
2               0       0    0   ...                       0.05   
3               0       0    0   ...                       0.00   
4               0       0    0   ...                       0.00   

    dst_host_same_src_port_rate  dst_host_srv_diff_host_rate  \
0                          0.17                         0.00   
1                          0.88                         0.00   
2                          0.00                         0.00   
3                          0.03                         0.04   
4                          0.00                         0.00   

   dst_host_serror_rate  dst_host_srv_serror_rate  dst_host_rerror_rate  \
0                  0.00                      0.00                  0.05   
1                  0.00                      0.00                  0.00   
2                  1.00                      1.00                  0.00   
3                  0.03                      0.01                  0.00   
4                  0.00                      0.00                  0.00   

   dst_host_srv_rerror_rate  attack_type  other    label  
0                      0.00       normal     20   normal  
1                      0.00       normal     15   normal  
2                      0.00      neptune     19  neptune  
3                      0.01       normal     21   normal  
4                      0.00       normal     21   normal  

[5 rows x 44 columns]

In [8]:
# DoS attacks
full . loc [ full . label == 'neptune', 'label'] = 'dos '
full . loc [ full . label == 'back', 'label'] = 'dos'
full . loc [ full . label == 'land', 'label'] = 'dos'
full . loc [ full . label == 'pod', 'label'] = 'dos'
full . loc [ full . label == 'smurf', 'label'] = 'dos'
full . loc [ full . label == 'teardrop', 'label'] = 'dos'
full . loc [ full . label == 'mailbomb', 'label'] = 'dos'
full . loc [ full . label == 'processtable', 'label'] = 'dos'
full . loc [ full . label == 'udpstorm', 'label'] = 'dos'
full . loc [ full . label == 'apache2', 'label'] = 'dos'
full . loc [ full . label == 'worm', 'label'] = 'dos'

# User -to - Root ( U2R )
full . loc [ full . label == 'buffer_overflow', 'label'] = 'u2r'
full . loc [ full . label == 'loadmodule', 'label'] = 'u2r'
full . loc [ full . label == 'perl', 'label'] = 'u2r'
full . loc [ full . label == 'rootkit', 'label'] = 'u2r'
full . loc [ full . label == 'sqlattack', 'label'] = 'u2r'
full . loc [ full . label == 'xterm', 'label'] = 'u2r'
full . loc [ full . label == 'ps', 'label'] = 'u2r'

# Remote -to - Local ( R2L)
full . loc [ full . label == 'ftp_write', 'label'] = 'r2l'
full . loc [ full . label == 'guess_passwd', 'label'] = 'r2l'
full . loc [ full . label == 'imap', 'label'] = 'r2l'
full . loc [ full . label == 'multihop', 'label'] = 'r2l'
full . loc [ full . label == 'phf', 'label'] = 'r2l'
full . loc [ full . label == 'spy', 'label'] = 'r2l'
full . loc [ full . label == 'warezclient', 'label'] = 'r2l'
full . loc [ full . label == 'warezmaster', 'label'] = 'r2l'
full . loc [ full . label == 'xlock', 'label'] = 'r2l'
full . loc [ full . label == 'xsnoop', 'label'] = 'r2l'
full . loc [ full . label == 'snmpgetattack', 'label'] = 'r2l'
full . loc [ full . label == 'httptunnel', 'label'] = 'r2l'
full . loc [ full . label == 'snmpguess', 'label'] = 'r2l'
full . loc [ full . label == 'sendmail', 'label'] = 'r2l'
full . loc [ full . label == 'named', 'label'] = 'r2l'

# Probe attacls
full . loc [ full . label == 'satan', 'label'] = 'probe'
full . loc [ full . label == 'ipsweep', 'label'] = 'probe'
full . loc [ full . label == 'nmap', 'label'] = 'probe'
full . loc [ full . label == 'portsweep', 'label'] = 'probe'
full . loc [ full . label == 'saint', 'label'] = 'probe'
full . loc [ full . label == 'mscan', 'label'] = 'probe'

full = full . drop ([ 'other', 'attack_type'], axis =1)
print ("Unique labels ", full . label . unique ())

Unique labels  ['normal' 'dos ' 'r2l' 'probe' 'dos' 'u2r']


In [9]:
full.head()

duration protocol   service flag  src_bytes  dst_bytes  land  \
0         0      tcp  ftp_data   SF        491          0     0   
1         0      udp     other   SF        146          0     0   
2         0      tcp   private   S0          0          0     0   
3         0      tcp      http   SF        232       8153     0   
4         0      tcp      http   SF        199        420     0   

   wrong_fragment  urgent  hot   ...    dst_host_srv_count  \
0               0       0    0   ...                    25   
1               0       0    0   ...                     1   
2               0       0    0   ...                    26   
3               0       0    0   ...                   255   
4               0       0    0   ...                   255   

   dst_host_same_srv_rate  dst_host_diff_srv_rate  \
0                    0.17                    0.03   
1                    0.00                    0.60   
2                    0.10                    0.05   
3                    1.00                    0.00   
4                    1.00                    0.00   

    dst_host_same_src_port_rate  dst_host_srv_diff_host_rate  \
0                          0.17                         0.00   
1                          0.88                         0.00   
2                          0.00                         0.00   
3                          0.03                         0.04   
4                          0.00                         0.00   

   dst_host_serror_rate  dst_host_srv_serror_rate  dst_host_rerror_rate  \
0                  0.00                      0.00                  0.05   
1                  0.00                      0.00                  0.00   
2                  1.00                      1.00                  0.00   
3                  0.03                      0.01                  0.00   
4                  0.00                      0.00                  0.00   

   dst_host_srv_rerror_rate   label  
0                      0.00  normal  
1                      0.00  normal  
2                      0.00    dos   
3                      0.01  normal  
4                      0.00  normal  

[5 rows x 42 columns]

In [10]:
# Generate One -Hot encoding
full2 = pd. get_dummies (full , drop_first = False )

In [11]:
full2.head()

duration  src_bytes  dst_bytes  land  wrong_fragment  urgent  hot  \
0         0        491          0     0               0       0    0   
1         0        146          0     0               0       0    0   
2         0          0          0     0               0       0    0   
3         0        232       8153     0               0       0    0   
4         0        199        420     0               0       0    0   

   num_failed_logins  logged_in  num_compromised    ...      flag_S2  flag_S3  \
0                  0          0                0    ...            0        0   
1                  0          0                0    ...            0        0   
2                  0          0                0    ...            0        0   
3                  0          1                0    ...            0        0   
4                  0          1                0    ...            0        0   

   flag_SF  flag_SH  label_dos  label_dos   label_normal  label_probe  \
0        1        0          0           0             1            0   
1        1        0          0           0             1            0   
2        0        0          0           1             0            0   
3        1        0          0           0             1            0   
4        1        0          0           0             1            0   

   label_r2l  label_u2r  
0          0          0  
1          0          0  
2          0          0  
3          0          0  
4          0          0  

[5 rows x 128 columns]

In [12]:
full2['label_dos'] = full2.iloc[:,-5]+full2.iloc[:,-6]
full2.drop(full2.columns[-6], axis=1)

duration  src_bytes  dst_bytes  land  wrong_fragment  urgent  hot  \
0             0        491          0     0               0       0    0   
1             0        146          0     0               0       0    0   
2             0          0          0     0               0       0    0   
3             0        232       8153     0               0       0    0   
4             0        199        420     0               0       0    0   
5             0          0          0     0               0       0    0   
6             0          0          0     0               0       0    0   
7             0          0          0     0               0       0    0   
8             0          0          0     0               0       0    0   
9             0          0          0     0               0       0    0   
10            0          0          0     0               0       0    0   
11            0          0          0     0               0       0    0   
12            0        287       2251     0               0       0    0   
13            0        334          0     0               0       0    0   
14            0          0          0     0               0       0    0   
15            0          0          0     0               0       0    0   
16            0        300      13788     0               0       0    0   
17            0         18          0     0               0       0    0   
18            0        233        616     0               0       0    0   
19            0        343       1178     0               0       0    0   
20            0          0          0     0               0       0    0   
21            0          0          0     0               0       0    0   
22            0        253      11905     0               0       0    0   
23         5607        147        105     0               0       0    0   
24            0          0          0     0               0       0    0   
25          507        437      14421     0               0       0    0   
26            0          0          0     0               0       0    0   
27            0        227       6588     0               0       0    0   
28            0        215      10499     0               0       0    0   
29            0        241       1400     0               0       0    0   
...         ...        ...        ...   ...             ...     ...  ...   
22514         0        115          0     0               0       0    0   
22515         0          0          0     0               0       0    0   
22516         0        322        396     0               0       0    0   
22517      7498          0         44     0               0       0    0   
22518         0        295        757     0               0       0    0   
22519      8209          0         15     0               0       0    0   
22520         0          0          0     0               0       0    0   
22521         0          1          1     0               0       0    0   
22522         0      54540       8314     0               0       0    2   
22523         0        289       9522     0               0       0    0   
22524        15         45        214     0               0       0    0   
22525      2064      55744          0     0               0       0    0   
22526         0        169       4997     0               0       0    0   
22527         0        236      16257     0               0       0    0   
22528         0       1032          0     0               0       0    0   
22529         0          9        139     0               0       0    0   
22530         0          0          0     0               0       0    0   
22531         0          0          0     0               0       0    0   
22532         0        264      14839     0               0       0    0   
22533         0        274       1623     0               0       0    0   
22534         0          0          0     0            

In [13]:
# Separate training and test sets again
features = list ( full2 . columns [: -5])
y_train = np. array ( full2 [0: df. shape [0]][[ 'label_normal', 'label_dos', 'label_probe', 'label_r2l', 'label_u2r']])
X_train = full2 [0: df. shape [0]][ features ]
y_test = np. array ( full2 [df. shape [0]:][[ 'label_normal', 'label_dos', 'label_probe', 'label_r2l', 'label_u2r']])
X_test = full2 [df. shape [0]:][ features ]

In [14]:
# Scale data
scaler = MinMaxScaler ().fit( X_train )
X_train_scaled = np. array ( scaler . transform ( X_train ))
X_test_scaled = np. array ( scaler . transform ( X_test ))

In [15]:
# Generate label encoding for Logistic regression
labels = full . label . unique ()
le = LabelEncoder ()
le. fit ( labels )
y_full = le. transform ( full . label )
y_train_l = y_full [0: df. shape [0]]
y_test_l = y_full [df. shape [0]:]

In [16]:
print (" Training dataset shape ", X_train_scaled .shape , y_train . shape )
print (" Test dataset shape ", X_test_scaled .shape , y_test . shape )
print (" Label encoder y shape ", y_train_l .shape , y_test_l . shape )

 Training dataset shape  (125973, 123) (125973, 5)
 Test dataset shape  (22544, 123) (22544, 5)
 Label encoder y shape  (125973,) (22544,)


In [17]:
X_train_scaled.shape

(125973, 123)

In [18]:
y_train

array([[1, 0, 0, 0, 0],
       [1, 0, 0, 0, 0],
       [0, 1, 0, 0, 0],
       ..., 
       [1, 0, 0, 0, 0],
       [0, 1, 0, 0, 0],
       [1, 0, 0, 0, 0]], dtype=uint8)

In [19]:
from keras import backend as K
from keras.models import Sequential, load_model
from keras.layers import Dense , Dropout
from keras.optimizers import RMSprop , adam

from cleverhans.attacks import SaliencyMapMethod 
from cleverhans.utils_tf import model_train , model_eval , batch_eval, model_argmax
from cleverhans.attacks_tf import jacobian_graph
from cleverhans.utils import other_classes, set_log_level

import tensorflow as tf
from tensorflow.python.platform import flags

import logging
import random

Using TensorFlow backend.


In [20]:
K.set_learning_phase(1)
set_log_level(logging.DEBUG)

In [21]:
def mlp_model ():
    model = Sequential()
    model.add(Dense (256, activation ='relu', input_shape =(X_train_scaled.shape[1],)))
    model.add(Dropout(0.4))
    model.add(Dense(256, activation ='relu'))
    model.add(Dropout(0.4))
    model.add(Dense(5, activation ='softmax'))
    model.compile(loss ='categorical_crossentropy', optimizer ='adam', metrics = ['accuracy'])

    model.summary()

    return model

In [22]:
def evaluate ():
    eval_params = {'batch_size': 128}
    accuracy = model_eval (sess, x, y, predictions, X_test_scaled, y_test, args = eval_params)
    print ('Test accuracy on legitimate test examples : ' + str(accuracy))

In [23]:
tf.set_random_seed(42)
sess = tf.Session()
print('Tensorflow session starts.')

x = tf.placeholder(tf.float32, shape = (None , X_train_scaled.shape[1]) )
y = tf.placeholder(tf.float32, shape = (None , 5 ))

init = tf.global_variables_initializer()
sess.run(init)

Tensorflow session starts.


In [24]:
model = mlp_model()
predictions = model(x)

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_1 (Dense)              (None, 256)               31744     
_________________________________________________________________
dropout_1 (Dropout)          (None, 256)               0         
_________________________________________________________________
dense_2 (Dense)              (None, 256)               65792     
_________________________________________________________________
dropout_2 (Dropout)          (None, 256)               0         
_________________________________________________________________
dense_3 (Dense)              (None, 5)                 1285      
Total params: 98,821
Trainable params: 98,821
Non-trainable params: 0
_________________________________________________________________


In [25]:
train_params = {'nb_epochs': 10, 'batch_size': 128, 'learning_rate': 0.01, 'verbose': 0}
model_train (sess, x, y, predictions, X_train_scaled, y_train, evaluate = evaluate, args = train_params)

[INFO 2018-04-25 15:00:10,991 cleverhans] Epoch 0 took 10.054657936096191 seconds
[DEBUG 2018-04-25 15:00:14,075 cleverhans] Batch 100


Test accuracy on legitimate test examples : 0.824920156139


[INFO 2018-04-25 15:00:29,471 cleverhans] Epoch 1 took 14.980695962905884 seconds
[DEBUG 2018-04-25 15:00:31,978 cleverhans] Batch 100


Test accuracy on legitimate test examples : 0.830110007097


[INFO 2018-04-25 15:00:43,460 cleverhans] Epoch 2 took 9.230578422546387 seconds
[DEBUG 2018-04-25 15:00:44,290 cleverhans] Batch 100


Test accuracy on legitimate test examples : 0.823411994322


[INFO 2018-04-25 15:00:59,237 cleverhans] Epoch 3 took 14.419797897338867 seconds
[DEBUG 2018-04-25 15:01:00,543 cleverhans] Batch 100


Test accuracy on legitimate test examples : 0.823899929028


[INFO 2018-04-25 15:01:10,442 cleverhans] Epoch 4 took 9.020922183990479 seconds
[DEBUG 2018-04-25 15:01:12,356 cleverhans] Batch 100


Test accuracy on legitimate test examples : 0.82855748758


[INFO 2018-04-25 15:01:21,557 cleverhans] Epoch 5 took 8.698034763336182 seconds
[DEBUG 2018-04-25 15:01:24,107 cleverhans] Batch 100


Test accuracy on legitimate test examples : 0.824520936835


[INFO 2018-04-25 15:01:37,458 cleverhans] Epoch 6 took 11.675309896469116 seconds
[DEBUG 2018-04-25 15:01:38,187 cleverhans] Batch 100


Test accuracy on legitimate test examples : 0.832194819021


[INFO 2018-04-25 15:01:47,580 cleverhans] Epoch 7 took 8.737219095230103 seconds
[DEBUG 2018-04-25 15:01:48,311 cleverhans] Batch 100


Test accuracy on legitimate test examples : 0.82585166785


[INFO 2018-04-25 15:01:58,427 cleverhans] Epoch 8 took 9.693399667739868 seconds
[DEBUG 2018-04-25 15:02:04,374 cleverhans] Batch 100


Test accuracy on legitimate test examples : 0.82767033357


[INFO 2018-04-25 15:02:20,937 cleverhans] Epoch 9 took 15.873785495758057 seconds
[DEBUG 2018-04-25 15:02:21,815 cleverhans] Batch 100
[INFO 2018-04-25 15:02:22,520 cleverhans] Completed model training.


Test accuracy on legitimate test examples : 0.82944464159


True

In [26]:
model_json = model.to_json()
with open("model.json", "w") as json_file:
    json_file.write(model_json)
model.save_weights("model.h5")
print("Saved model to disk")

Saved model to disk


In [28]:
from keras.models import model_from_json
json_file = open('model.json', 'r')
loaded_model_json = json_file.read()
json_file.close()
loaded_model = model_from_json(loaded_model_json)
loaded_model.load_weights("model.h5")
print("Loaded model from disk")

Loaded model from disk


In [29]:
X_test_scaled[0]

array([ 0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  0.        ,  0.4481409 ,
        0.01956947,  0.        ,  0.        ,  1.        ,  1.        ,
        0.04      ,  0.06      ,  0.        ,  1.        ,  0.03921569,
        0.04      ,  0.06      ,  0.        ,  0.        ,  0.        ,
        0.        ,  1.        ,  1.        ,  0.        ,  1.        ,
        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  0.        ,  0.  

In [33]:
np.save('X_train_scaled', X_train_scaled)
np.save('y_train', y_train)

source_samples = X_test_scaled.shape[0]

X_adv = np.empty((0,123))

jsma = SaliencyMapMethod(model, back = 'tf', sess = sess)
jsma_params = {'theta': 1., 'gamma': 0.1, 'clip_min': 0., 'clip_max': 1., 'y_target': None}

for sample_ind in range(0, source_samples):
    
    print('Generating sample no: ' + str(sample_ind))
    
    sample = X_test_scaled[sample_ind:(sample_ind+1)]
    
    current_class = int(np.argmax(y_test[sample_ind]))
    target_classes = other_classes(5, current_class)
    
    if current_class != 0 :
        target_classes = [0]
    else :
        rand_class = random.randint(1,4)
        target_classes = [rand_class]
        
    
    for target in target_classes:
        
        one_hot_target = np.zeros((1, 5), dtype=np.float32)
        one_hot_target[0, target] = 1
        jsma_params['y_target'] = one_hot_target
        
        adv_x = jsma.generate_np(sample, **jsma_params)
        
        
        X_adv = np.vstack([X_adv, adv_x])

print (X_adv.shape)

eval_params = {'batch_size': 128}
accuracy = model_eval(sess, x, y, predictions, X_test_scaled, y_test, args = eval_params)
print ('Test accuracy on normal examples : ' + str(accuracy))

accuracy_adv = model_eval(sess, x, y, predictions, X_adv , y_test, args = eval_params )
print ('Test accuracy on adversarial examples : ' + str(accuracy_adv))

np.save('Adv_samples_1', X_adv)

np.save('X_test_scaled', X_test_scaled)
np.save('y_test', y_test)